**Goldbach's conjecture.**
Proposed by Christian Goldbach in a letter to Euler in 1742, Goldbach's
conjecture states that every even integer greater than 2 can be expressed
as the sum of two prime numbers:

$$
\forall\, n \in \mathbb{Z},\; n > 2,\; n \text{ even}
\quad\Rightarrow\quad
\exists\, p, q \in \mathbb{P} : n = p + q
$$

For example: $4 = 2+2$, $6 = 3+3$, $8 = 3+5$, $100 = 3+97 = 11+89 = \ldots$
Despite being verified computationally for all even integers up to
$4 \times 10^{18}$, the conjecture remains unproven.
This notebook verifies it numerically for all even integers up to 900
by generating every pair of primes from a fixed set and checking
whether any even number in the range is missing from the sums.

In [1]:
"""goldbach_conjecture.ipynb"""

# Cell 01 - Build an array of the first 99 primes using SymPy

import numpy as np
from sympy import prime

test_limit = 99

# prime(k) returns the k-th prime (1-indexed): prime(1)=2, prime(2)=3, ...
primes = np.array([prime(n) for n in range(1, test_limit + 1)])

print(f"First 10 primes : {primes[:10]}")
print(f"Last 10 primes  : {primes[-10:]}")
print(f"Largest prime   : {primes[-1]}")

First 10 primes : [ 2  3  5  7 11 13 17 19 23 29]
Last 10 primes  : [463 467 479 487 491 499 503 509 521 523]
Largest prime   : 523


**Generating all prime pairs.**
`combinations_with_replacement` produces every unordered pair $(p, q)$
where $p \leq q$, including pairs where both primes are the same (e.g. $(2,2)$,
$(3,3)$).
With 99 primes this gives $\binom{99+1}{2} = 4{,}950$ unique pairs.
We allow $p = q$ because Goldbach permits the same prime to be used twice
- for example $4 = 2+2$ and $6 = 3+3$.

In [2]:
# Cell 02 - Generate all unordered pairs of primes (repetition allowed)

from itertools import combinations_with_replacement

prime_pairs = list(combinations_with_replacement([int(p) for p in primes], 2))

print(f"Number of unique prime pairs : {len(prime_pairs):,}")
print(f"First 5 pairs : {prime_pairs[:5]}")
print(f"Last 5 pairs  : {prime_pairs[-5:]}")

Number of unique prime pairs : 4,950
First 5 pairs : [(2, 2), (2, 3), (2, 5), (2, 7), (2, 11)]
Last 5 pairs  : [(509, 521), (509, 523), (521, 521), (521, 523), (523, 523)]


**Computing unique pairwise sums.**
Many different prime pairs share the same sum - for example
$100 = 3+97 = 11+89 = 17+83 = \ldots$ - so we take the unique
values only.
Note that the sums include both even and odd values: odd sums arise
from pairs involving 2 (the only even prime), such as $2+3=5$ or $2+5=7$.
Only the even sums are relevant to Goldbach's conjecture.

In [3]:
# Cell 03 - Compute the sorted unique sums of all prime pairs

summed_pairs = np.unique([p + q for p, q in prime_pairs])

print(f"Number of unique sums  : {len(summed_pairs):,}")
print(f"First 10 sums          : {summed_pairs[:10]}")
print(f"Last 10 sums           : {summed_pairs[-10:]}")
print(f"Largest sum (p[-1]*2)  : {primes[-1] * 2}")

Number of unique sums  : 611
First 10 sums          : [ 4  5  6  7  8  9 10 12 13 14]
Last 10 sums           : [1018 1020 1022 1024 1026 1030 1032 1042 1044 1046]
Largest sum (p[-1]*2)  : 1046


**Checking for violations.**
`np.setdiff1d(A, B)` returns elements present in `A` but not in `B`.
We use it to find any even integer in the range $[4, 900]$ that cannot
be expressed as a sum of two primes from our set.
Our 99 primes reach up to 523, giving a maximum sum of 1046, which
comfortably covers the entire verification range up to 900.
A result of zero violations confirms Goldbach's conjecture for all
even integers from 4 to 900.

In [4]:
# Cell 04 - Find any even integer in [4, 900] not expressible as a sum of two primes

even_range = np.arange(4, 902, 2)  # all even integers from 4 to 900 inclusive
violations = np.setdiff1d(even_range, summed_pairs)

print(f"Even integers checked          : {len(even_range):,}")
print(f"Goldbach violations up to 900  : {len(violations)}")
if len(violations) > 0:
    print(f"Violations : {violations}")

Even integers checked          : 449
Goldbach violations up to 900  : 0
